src/main/java/seedu/address/logic/newcommands/MarkCommand.java

```diff
@@ -0,0 +1,117 @@
+import seedu.address.logic.newcommands.exceptions.CommandException;
+import seedu.address.model.path.AbsolutePath;
+import seedu.address.model.statemanager.State;
+import seedu.address.model.statemanager.TaskOperation;
+import seedu.address.model.taskmanager.Task;
+
+/**
+ * The MarkCommand class represents a command for marking a task in the task list.
+ * This command is typically used in a context where the user is viewing a list of tasks.
+ * It marks a specific task in the list as completed.
+ */
+public class MarkCommand extends Command {
+
+    public static final String COMMAND_WORD = "mark";
+
+    public static final String MESSAGE_INCORRECT_STATE = "The current state is not showing task list.";
+
+    public static final String MESSAGE_INVALID_INDEX = "The task index provided is invalid.";
+
+    public static final String MESSAGE_MARK_TASK_SUCCESS = "Marked task: %1$s";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + " [task index]";
+
+    private static final Logger logger = LogsCenter.getLogger(MarkCommand.class);
+
+    private final Index index;
+
+    /**
+     * Constructs a MarkCommand with the specified task index to be marked.
+     *
+     * @param index The index of the task to be marked.
+     */
+    public MarkCommand(Index index) {
+        requireNonNull(index);
+        this.index = index;
+    }
+
+    /**
+     * Executes the MarkCommand to mark a task as completed.
+     *
+     * @param state The current state of the application.
+     * @return A CommandResult containing a message indicating the success of the marking operation.
+     * @throws CommandException If the command cannot be executed due to an incorrect state (not showing task list).
+     */
+    @Override
+    public CommandResult execute(State state) throws CommandException {
+        requireNonNull(state);
+
+        logger.info("Executing mark task command...");
+
+        if (state.isShowTaskList()) {
+            AbsolutePath displayPath = state.getDisplayPath();
+            TaskOperation taskOperation = state.taskOperation(displayPath);
+
+            // Check if index is valid.
+            if (!taskOperation.isValidIndex(this.index.getOneBased())) {
+                logger.warning("Invalid index: " + this.index.getOneBased() + ". Aborting mark task command.");
+                throw new CommandException(MESSAGE_INVALID_INDEX);
+            }
+
+            logger.info("Executing mark task command on index " + this.index.getOneBased());
+
+            Task markedTask = taskOperation.markTask(this.index.getOneBased());
+            state.updateList();
+
+            logger.info("Task marked successfully. Marked task: " + markedTask.toString());
+
+            return new CommandResult(String.format(MESSAGE_MARK_TASK_SUCCESS, markedTask));
+        } else {
+            logger.warning("Task list is not shown. Aborting mark task command.");
+            throw new CommandException(MESSAGE_INCORRECT_STATE);
+        }
```

Comment:
Consider using guard clause to replace nested condition. Refer to this https://nus-cs2103-ay2324s1.github.io/website/se-book-adapted/chapters/codeQuality.html#make-the-happy-path-prominent